blog: Mutation switching

nicojs opened the pull request.

nicojs commented:
@legopiraat I didn't read through it all, but is it correct to assume you're trying to reach 2 goals? 

1. Anouncing that we support C# and Scala in preview
2. Explaining what mutation switching is

Those to topics are great, but in my experience it is better to keep blogs focussed on 1 goal and keep them short and sweet. 

Could we split this in 2 articles? One to announce the new frameworks and what this means to Stryker for javascript and one article about mutation switching?

richardwerkman commented:
I think the blog is great! It's goal is clear to me, it explains how Stryker4s implements mutation switching and why we choose this. We could also dedicate another blog to how Stryker.NET uses mutation switching?

legopiraat commented:
Sounds like a plan 👍 Any furious blogger that wants to write a warm welcome blog?

nicojs commented:
@legopiraat @richardwerkman 

See #94

hugo-vrijswijk commented on src/blog/2018-09-24/mutation-switching.pug:
````diff
@@ -0,0 +1,113 @@
+             }
+         }
+        ```
+        If we follow the transform, compile, test and collect the results step base approach we would have to compile our code three times. 
````
This reads like the 3 times compilation is because of the steps, not because of the 3 mutations

legopiraat replied:
Is this better?

hugo-vrijswijk commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,113 @@
+        ## Mutation switching
+        Mutation switching to the rescue! So how is mutation switching faster than compiling each mutation on its own? 
+        The steps are quite similar but have some minor differences. First, All possible mutations are identified for the whole code base. 
+        Secondly, all mutations are applied to the code base using a [Scala Pattern match](https://docs.scala-lang.org/tour/pattern-matching.html). 
```
... with the original statement in the default case, and activated using environment variables

legopiraat replied:
added :)

legopiraat commented on src/blog/2018-09-24/mutation-switching.pug:
````diff
@@ -0,0 +1,113 @@
+        One way of introducing mutations to a codebase is mutating one statement, compile to code, run tests and gather the results. 
+        This is a logical choice because you have the maximum flexibility and are certain what effect the mutant had. 
+        When looking at the following example there are three mutations possible, we can change the `>` into `>=`, `<` and  `==`.
+        ```scala
````
done

legopiraat commented on src/blog/2018-09-24/blog.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "title": "Mutation switching",
+    "description": "How does Stryker4s reduce compilation times using mutation switching",
```
fixed

kmdrGroch commented:
@legopiraat in file ``src/blog/2018-09-24/blog.json``
you have written:
``"description": "How Stryker4s and Stryker.Net use mutation switching to reduce compilation times.",``

Well. It should be a question and question ends with ``?`` usually.
And there is no inversion at all.

And there should be ``time`` instead of ``times``
you can multiply using ``times``, but besides it i think ``time`` has no plural form (i can be wrong, then tell me) [dont take it as an offence pls]

so I suggest:
``"description": "How do Stryker4s and Stryker.Net use mutation switching to reduce compilation time?",``

legopiraat commented:
Fixed the issues right away :)

legopiraat commented:
@kmdrGroch this time I didn't 🙈

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+    :markdown-it
+        
+        We are very excited because Stryker has a few new friends in the Stryker landscape! One of those new friends is Stryker4s(cala). 
+        With Stryker now being available for Scala, we opened up a whole new landscape for people to get started with mutation testing. 
```
You're reusing the word "landscape" here. Sounds redundant.

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        With Stryker now being available for Scala, we opened up a whole new landscape for people to get started with mutation testing. 
+        The adaptation of Scala gave us a lot of joy but also a lot of new problems. One of those problems was the Scala compiler, 
+        we all know it's not the [fastest of its kind](https://scala-ci.typesafe.com/grafana/dashboard/db/scala-benchmark?from=1429139130109&to=1492531027936h&orgId=1&var-branch=2.11.x&var-branch=2.12.x&var-branch=2.13.x&var-source=scala&var-bench=HotScalacBenchmark.compile&var-host=scalabench@scalabench@) 
+        and many many hours are being put in to make it as fast as possible. With the goals of stryker in mind (It's easy to use and fast to run) 
```
Capital letter S for Stryker :)

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        and many many hours are being put in to make it as fast as possible. With the goals of stryker in mind (It's easy to use and fast to run) 
+        we needed to come up with an intelligent way to deal with compilation time.
+        
+        ## Compiling each mutation
```
To make sure people understand that compiling each mutant is _not_ what we're doing, we should rename this chapter. How about:

> Attempt one: Compiling each mutant

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        If we assume the compile time of this program is 10 seconds, we already have 30 seconds of compile time for one full mutation run. 
+        This would quickly get out of hand when the code base is bigger and has more mutations available.
+        
+        ## Mutation switching
```
You could also add a small chapter about **Attempt 2: mutating bytecode**. In there you can explain that this does not scale for scala, with all the minor releases breaking the runtime contracts.

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        Mutation switching to the rescue! So how is mutation switching faster than compiling each mutation on its own? 
+        The steps are quite similar but have some minor differences. First, All possible mutations are identified for the whole code base. 
+        Secondly, all mutations are applied to the code base using a [Scala Pattern match](https://docs.scala-lang.org/tour/pattern-matching.html) 
+        with the original satement in the default case, and activated using environment variables.
```
typo: statement.

Also, this is a long sentence. Can we shorten it? Maybe move out the "and activated using environment variables" part to a new sentence?

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
````diff
@@ -0,0 +1,116 @@
+        ```
+        All possible mutations are implemented in the pattern match with an identifier so we can turn on specific mutations 
+        and a default case so the code will work normally one none of the mutants are activated. Now the code base only needs to be compiled once, 
+        but because of the increased size of the code base the compilation time will be a bit longer but that does not way up 
````
Suggestion: "but that does not way up against the full recompilation as before" -> "but that is negligible compared to the overhead of compiling each mutant."

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        but because of the increased size of the code base the compilation time will be a bit longer but that does not way up 
+        against the full recompilation as before. If the compilation time for the code base would be 15 seconds we will still 
+        gain 15 seconds which is already a significant performance booster. So we gained a lot of performance without losing any 
+        flexibility because of the ability to turn mutations on and off using the environment variable. 
```
I'm missing a comma before "because"

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+            case _ => .filterNot(isOdd)
+          }).
+          (sys.env.get("ACTIVE_MUTATION") match {
+            case Some("1") =>  .filterNot(_.equals(specialNumber))
```
Shouldn't this be `Some("2")`?

legopiraat replied:
To be even more correct the first `Some("1")` should be `Some("0")` but good catch

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        flexibility because of the ability to turn mutations on and off using the environment variable. 
+        
+        ## Top statements
+        With mutation switching, we reduced compilation times by a lot but soon noticed there was one more problem to tackle. 
```
This introduction text is petty long. How about making it short and sweet? For example:

"Mutation switching sure is great, but let's look at a more complex example."

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
````diff
@@ -0,0 +1,116 @@
+        }
+        ```
+        As you can see it's already a lot more complicated and the code doesn't even compile! 
+        To reduce a lot of the complexity and make the code actually compile we need to take a deeper look in the abstract syntax tree 
````
I would leave out the "complexity" part. It makes the text opinionated, let the reader decide for itself if it is complex or not. The most important take away is that the code doesn't compile, right?

nicojs commented on src/blog/2018-09-24/mutation-switching.pug:
```diff
@@ -0,0 +1,116 @@
+        This enables us to get clean, readable pattern matches and avoid compilation errors like we've seen before. 
+        
+        ## Conclusion
+        With the combination of mutation switching and traversing back to the top statements Stryker4s is able to apply mutations to the codebase 
```
What is the AST term for a "top statement"? Can we use that term instead of "top statement"?

legopiraat replied:
I think its just statement i will change it to parent statement i guess thats clearer

nicojs commented:
@legopiraat  thanks for updating! I took the liberty of streamlining some parts, hope you didn't mind. We can still change it back if you want, just thought these changes would make the text a little more active and engaging.

I've also added a "what's next" part, inviting people to join.

Please let me know what you think, or change anything you don't like.

legopiraat commented:
Today or tomorrow is fine :) not in a hurry

nicojs commented:
Ok, I'll read it tomorrow after a full night of sleep. Probably big typo's lurking in there 😅

nicojs commented:
I've fixed some typo's and updated the date to today. I also removed C# from the article. It got only mentioned once. I thought this was somewhat confusing. Better to write an article to focus on C# later.

@hugo-vrijswijk @legopiraat could you guys take one final look?

richardwerkman commented:
@nicojs I've already started on a C# specific blogpost. This post is real nice!

legopiraat commented on src/blog/2018-10-5/mutation-switching.pug:
```diff
@@ -0,0 +1,161 @@
+        
+        ## Attempt one: Compiling each mutation
+        
+        One way of introducing mutants to a codebase is mutating one statement, compile to code, run tests, gather the results and repeat. 
```
Shouldn't this be ` compile the code, run the tests,`  instead of `compile to code, run tests`?

nicojs replied:
Good find!

nicojs commented:
I've merged it. Any typo's left will need to be fixed afterwards 🙈
